ggml-vulkan-shaders.hpp

```diff
@@ -0,0 +1,773 @@
+#include <string>
+
+// Generic
+const std::string shader_f32 = R"(
```

Comment:
is there a reason to use `std::string` here and not just `const char*` ?

Reply:
The Vulkan code is C++, so I used string here instead of const char. But it seems that compiling shaders on runtime causes more issues than it resolves. It seems to cause issues with building on Windows and also @Jipok 's error was with linking shaderc. I'll replace it with something that builds and compiles the shaders on build time, not run time.

Reply:
> The Vulkan code is C++, so I used string here instead of const char.

yea, but no reason to construct a `std::string`

> But it seems that compiling shaders on runtime causes more issues than it resolves.

yea i feel that. guess we will have to ship optimized `.spv` files.

Reply:
I'm not familar with Vulkan but are the shaders specially optimized for the local hardware when they're compiled? This would be a strong argument for runtime compilation as it would make the builds portable across different platforms.

Reply:
That was the idea. Currently it only distinguishes between devices with shaderFloat16 and devices without it, which would be easy to ship with SPIR-V files. Another use case will be tensor core usage with NV_cooperative_matrix support. There's probably not enough cases out there to warrant dealing with the issues linking shaderc is causing.

Reply:
If you mean automatic optimization for the specific device, that still happens when the driver compiles from the intermediate binary format SPIR-V (.spv file) to whatever it's using internally. I was talking about compiling GLSL code on runtime to SPIR-V.

Reply:
There's also [VK_KHR_cooperative_matrix](https://github.com/KhronosGroup/SPIRV-Registry/pull/207)